File: AtlanticOsloM/OpenQVT.m
module OpenQVT {

	type Variable {
		Name_Att : Text;
		Type_Att : Text;
		Value_Att : Text;
		Visibility_Att : Text;
		Context_Ref : Context;
	};

	type Context {
		Variables_ContRef : Variable*;
		Rule_Ref : Rule;
	};

	type Rule {
		Context_ContRef : Context;
		SubRule_ContRef : Rule;
		SuperRule_Ref : Rule;
		Services_ContRef : Service*;
	};

	type RootRule : Rule {
		RuleSet_Ref : RuleSet;
	};

	type RuleSetCall : Rule {
		RuleSet_Ref : RuleSet;
	};

	type QARule : Rule {
		Query_ContRef : Query;
		Action_ContRef : Action;
	};

	type RuleSet {
		RootRule_ContRef : RootRule;
	};

	type Service {
		Visibility_Att : Text;
		Signature_Att : Text;
		Rule_Ref : Rule;
	};

	type OCLService : Service;

	type RTransService : Service;

	type JavaService : Service;

	type RulePart {
		Service_Ref : Service;
	};

	type Action : RulePart {
		QARule_Ref : QARule;
	};

	type RTransAction : Action;

	type JavaAction : Action;

	type Query : RulePart {
		QARule_Ref : QARule;
	};

	type OCLFilter : Query;

	type RTransQuery : Query;

	type JavaQuery : Query;

	type JavaFilter : Query;
}

File: AtlanticClojure/OpenQVT.clj
( defstruct Variable :name :type :value :visibility :context )
( defstruct Context :rule )
( defstruct Rule :context :subRule :superRule )
( defstruct RootRule :ruleSet :context :subRule :superRule )
( defstruct RuleSetCall :ruleSet :context :subRule :superRule )
( defstruct QARule :query :action :context :subRule :superRule )
( defstruct RuleSet :rootRule )
( defstruct Service :visibility :signature :rule )
( defstruct OCLService :visibility :signature :rule )
( defstruct RTransService :visibility :signature :rule )
( defstruct JavaService :visibility :signature :rule )
( defstruct RulePart :service )
( defstruct Action :qARule :service )
( defstruct RTransAction :qARule :service )
( defstruct JavaAction :qARule :service )
( defstruct Query :qARule :service )
( defstruct OCLFilter :qARule :service )
( defstruct RTransQuery :qARule :service )
( defstruct JavaQuery :qARule :service )
( defstruct JavaFilter :qARule :service )